artifact/sha/sha.cpp

```diff
@@ -0,0 +1,126 @@
+#include <artifact/sha/sha.hpp>
+
+#include <common/common.hpp>
+
+#include <cstdint>
+#include <string>
+#include <ctime>
+#include <iomanip>
+#include <vector>
+#include <sstream>
+#include <algorithm>
+
+namespace mender {
+namespace sha {
+
+namespace log = mender::common::log;
+
+
+const ErrorCategoryClass ErrorCategory = ErrorCategoryClass();
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "ShaSumErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	InitializationError:
+		return "Initialization error";
+	ShasumCreationError:
+		return "Shasum creation error";
+	ShasumMismatchError:
+		return "Shasum mismatch error";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code) {
+	return error::Error(error_condition(code, ErrorCategory), "");
+}
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+Reader::Reader(io::Reader &reader) :
+	Reader::Reader {reader, ""} {
+}
+
+Reader::Reader(io::Reader &reader, std::string expected_sha = "") :
```

Comment:
&expected_sha

Reply:
Yes, but use `const` if you use reference (unless you intend to modify it).